libraries/stdlib/src/kotlin/io/Console.kt

```diff
@@ -132,50 +136,91 @@ public inline fun println() {
     System.out.println()
 }
 
-// Since System.in can change its value on the course of program running,
-// we should always delegate to current value and cannot just pass it to InputStreamReader constructor.
-// We could use "by" implementation, but we can only use "by" with interfaces and InputStream is abstract class.
-private val stdin: BufferedReader by lazy { BufferedReader(InputStreamReader(object : InputStream() {
-    public override fun read(): Int {
-        return System.`in`.read()
-    }
+private const val BUFFER_SIZE: Int = 32
+private const val LINE_SEPARATOR_MAX_LENGTH: Int = 2
 
-    public override fun reset() {
-        System.`in`.reset()
-    }
+private var cachedDecoder: CharsetDecoder? = null
 
-    public override fun read(b: ByteArray): Int {
-        return System.`in`.read(b)
+private fun decoderFor(charset: Charset): CharsetDecoder {
+    return cachedDecoder?.takeIf { cached ->
+        cached.charset() == charset
+    } ?: charset.newDecoder().also { newDecoder ->
+        cachedDecoder = newDecoder
     }
+}
 
-    public override fun close() {
-        System.`in`.close()
+/**
+ * Reads a line of input from the standard input stream.
+ *
+ * @return the line read or `null` if the input stream is redirected to a file and the end of file has been reached.
+ */
+fun readLine(charset: Charset = Charset.defaultCharset()): String? = readLine(System.`in`, decoderFor(charset))
+
+internal fun readLine(inputStream: InputStream, decoder: CharsetDecoder): String? {
+    require(decoder.maxCharsPerByte() <= 1) { "Encodings with multiple chars per byte are not supported" }
+
+    val byteBuffer = ByteBuffer.allocate(BUFFER_SIZE)
+    val charBuffer = CharBuffer.allocate(LINE_SEPARATOR_MAX_LENGTH)
+    val stringBuilder = StringBuilder()
+
+    var read = inputStream.read()
+    if (read == -1) return null
+    while (read != -1) {
```

Comment:
Always true, this check can be moved to the end of loop, just replace `while () { }` with `do { } while ()`